crates/jxl-frame/src/data/spline.rs

```diff
@@ -1,26 +1,371 @@
+    fmt::Display,
+    io::Read,
+    ops::{Add, Mul, Sub},
+};
 
-use jxl_bitstream::{Bitstream, Bundle};
+use jxl_bitstream::{unpack_signed, Bitstream, Bundle};
+use jxl_coding::Decoder;
 
-use crate::Result;
+use crate::{FrameHeader, Result};
 
+const MAX_NUM_SPLINES: usize = 1 << 24;
+const MAX_NUM_CONTROL_POINTS: usize = 1 << 20;
+
+/// Holds quantized splines
 #[derive(Debug)]
 pub struct Splines {
-    coords: Vec<(i32, i32)>,
-    quant_adjust: i32,
+    pub quant_splines: Vec<QuantSpline>,
+    pub quant_adjust: i32,
+}
+
+/// 2D Point in f32 coordinates
+#[derive(Debug, Default, Clone, Copy)]
+pub struct Point {
+    pub x: f32,
+    pub y: f32,
+}
+
+/// Holds control point coordinates and dequantized DCT32 coefficients of XYB channels, σ parameter of the spline
+#[derive(Debug)]
+pub struct Spline {
+    pub points: Vec<Point>,
+    pub xyb_dct: [[f32; 32]; 3],
+    pub sigma_dct: [f32; 32],
+}
+
+pub struct SplineArc {
+    pub point: Point,
+    pub lenght: f32,
+}
+
+/// Holds delta-endcoded control points coordinates (without starting point) and quantized DCT32 coefficients
+///
+/// Use [`QuantSpline::dequant`] to get normal [Spline]
+#[derive(Debug, Default, Clone)]
+pub struct QuantSpline {
+    start_point: (i32, i32),
+    points_deltas: Vec<(i32, i32)>,
+    xyb_dct: [[i32; 32]; 3],
+    sigma_dct: [i32; 32],
 }
 
-impl<Ctx> Bundle<Ctx> for Splines {
+impl Bundle<&FrameHeader> for Splines {
     type Error = crate::Error;
 
-    fn parse<R: Read>(bitstream: &mut Bitstream<R>, _: Ctx) -> Result<Self> {
+    fn parse<R: Read>(bitstream: &mut Bitstream<R>, header: &FrameHeader) -> Result<Self> {
         let mut decoder = jxl_coding::Decoder::parse(bitstream, 6)?;
         decoder.begin(bitstream)?;
+        let num_pixels = (header.width * header.height) as usize;
+
+        let num_splines = (decoder.read_varint(bitstream, 2)? + 1) as usize;
+
+        let max_num_splines = usize::min(MAX_NUM_SPLINES, num_pixels / 4);
+        if num_splines > max_num_splines {
+            return Err(crate::Error::TooManySplines(num_splines));
+        }
+
+        let mut start_points = vec![(0i32, 0i32); num_splines];
+        for i in 0..num_splines {
+            let mut x = decoder.read_varint(bitstream, 1)? as i32;
+            let mut y = decoder.read_varint(bitstream, 1)? as i32;
+            if i != 0 {
+                x = unpack_signed(x as u32) + start_points[i - 1].0;
+                y = unpack_signed(y as u32) + start_points[i - 1].1;
+            }
+            start_points[i].0 = x;
+            start_points[i].1 = y;
+        }
+
+        let quant_adjust = unpack_signed(decoder.read_varint(bitstream, 0)?);
+
+        let mut splines: Vec<QuantSpline> = Vec::with_capacity(num_splines);
+        for start_point in start_points {
+            let mut spline = QuantSpline::new(start_point);
+            spline.decode(&mut decoder, bitstream, num_pixels)?;
+            spline.start_point = start_point;
```

Comment:
`QuantSpline::new` does this, so this can be removed.

Reply:
Whoops